### ISSUE TITLE
Issue with jupyterhub image gcr.io/kubeflow-images-staging/tensorflow-notebook-cpu

### ISSUE THREAD

**Issue by pvellanki** [kubeflow/kubeflow#554]
Kubeflow is setup on-prem with external nfs server as storage

When Jupyter notebook is started, notebook gets created using the following image gcr.io/kubeflow-images-staging/tensorflow-notebook-cpu but it gets deleted after 30 seconds.  Please see below the status of the POD as images gets pulled, gets created and then deleted in 30s

Any ideas as to how to debug this...

Events:
  Type     Reason                 Age                From               Message
  ----     ------                 ----               ----               -------
  Warning  FailedScheduling       56s (x2 over 56s)  default-scheduler  pod has unbound PersistentVolumeClaims
  Normal   Scheduled              55s                default-scheduler  Successfully assigned jupyter-prasad to kubevm2
  Normal   SuccessfulMountVolume  54s                kubelet, kubevm2   MountVolume.SetUp succeeded for volume "no-api-access-please"
  Normal   SuccessfulMountVolume  54s                kubelet, kubevm2   MountVolume.SetUp succeeded for volume "pvc-33aa2fe5-3579-11e8-bcc7-52540067611f"
  Normal   Pulling                52s                kubelet, kubevm2   pulling image "gcr.io/kubeflow-images-staging/tensorflow-notebook-cpu"

Events:
  Type     Reason                 Age              From               Message
  ----     ------                 ----             ----               -------
  Warning  FailedScheduling       6m (x2 over 6m)  default-scheduler  pod has unbound PersistentVolumeClaims
  Normal   Scheduled              6m               default-scheduler  Successfully assigned jupyter-prasad to kubevm2
  Normal   SuccessfulMountVolume  6m               kubelet, kubevm2   MountVolume.SetUp succeeded for volume "no-api-access-please"
  Normal   SuccessfulMountVolume  6m               kubelet, kubevm2   MountVolume.SetUp succeeded for volume "pvc-33aa2fe5-3579-11e8-bcc7-52540067611f"
  Normal   Pulling                6m               kubelet, kubevm2   pulling image "gcr.io/kubeflow-images-staging/tensorflow-notebook-cpu"
  Normal   Pulled                 1s               kubelet, kubevm2   Successfully pulled image "gcr.io/kubeflow-images-staging/tensorflow-notebook-cpu"
  Normal   Created                1s               kubelet, kubevm2   Created container
  Normal   Started                1s               kubelet, kubevm2   Started container

  Events:
  Type     Reason                 Age              From               Message
  ----     ------                 ----             ----               -------
  Warning  FailedScheduling       7m (x2 over 7m)  default-scheduler  pod has unbound PersistentVolumeClaims
  Normal   Scheduled              7m               default-scheduler  Successfully assigned jupyter-prasad to kubevm2
  Normal   SuccessfulMountVolume  7m               kubelet, kubevm2   MountVolume.SetUp succeeded for volume "no-api-access-please"
  Normal   SuccessfulMountVolume  7m               kubelet, kubevm2   MountVolume.SetUp succeeded for volume "pvc-33aa2fe5-3579-11e8-bcc7-52540067611f"
  Normal   Pulling                7m               kubelet, kubevm2   pulling image "gcr.io/kubeflow-images-staging/tensorflow-notebook-cpu"
  Normal   Pulled                 38s              kubelet, kubevm2   Successfully pulled image "gcr.io/kubeflow-images-staging/tensorflow-notebook-cpu"
  Normal   Created                38s              kubelet, kubevm2   Created container
  Normal   Started                38s              kubelet, kubevm2   Started container
  Normal   Killing                3s               kubelet, kubevm2   Killing container with id docker://notebook:Need to kill Pod

![screen shot 2018-04-01 at 12 00 00 am](https://user-images.githubusercontent.com/8160576/38170701-b5e8a94e-3540-11e8-8119-eaedeaab6886.png)

**Comment by pdmack** [comment]
`Successfully assigned jupyter-prasad to kubevm2`

Is it possible your vSphere environment doesn't have the necessary resources to run the notebook pod (i.e., CPU, memory)? What cpu_guarantee and mem_guarantee are you requesting for your notebook?

Docker has been a little notorious in the past about swallowing up OOM-type events and messages.

**Comment by pvellanki** [comment]
Am running on a KVM based VMs. Each VM has 6 vCPUs and 8G. I was giving 2Gi memory and 1 CPU for NB. 

What is the best way to debug this. Am building local NB docker (still facing the same issue) to put some debug prints. It looks like the NB server is not available at port 8888 which JH tries to talk to.

**Comment by pdmack** [comment]
Can you make sure that image runs standalone in your Docker env?

`docker run -it --rm --user root -e GRANT_SUDO=yes gcr.io/kubeflow-images-staging/tensorflow-notebook-cpu:latest`

If you don't see `The Jupyter Notebook is running at: ...` then the error will hopefully be informative.

**Comment by pvellanki** [comment]
@pdmack - Thanks for the help
Looks like jupyter is exiting with the following

**[I 01:21:23.843 NotebookApp] Writing notebook server cookie secret to /home/jovyan/.local/share/jupyter/runtime/notebook_cookie_secret
[W 01:21:24.528 NotebookApp] WARNING: The notebook server is listening on all IP addresses and not using encryption. This is not recommended.
Illegal instruction (core dumped)**

Tested this with various hardware it still happens..

Built docker with base image from jupyter github https://github.com/jupyter/docker-stacks and included keras, tensorflow and keras. That works. 
Something in the image on gcr does not sit well with the systems (ubuntu16.04) we have here. Not sure if others faced this issue

**Comment by pvellanki** [comment]
Yup.  some of the machines I was testing was westmere based VMs. 
Will test with post westmere and also KVM based VMs on skylake and post westmere

**Comment by pvellanki** [comment]
as @pdmack said, it looks like KVM also has the same issue with this. Tested on plain KVM VM. 
In which case, This would be an issue on most on prem environments.

**Comment by myqq0000** [comment]
The same issue in my own server... Follow the latest [user_guide](https://github.com/kubeflow/kubeflow/blob/master/user_guide.md).
I have tried it 5 days ago. Because of the PVC problem can't set up JupyterHub.
And today I found the document is updated. So I tried it again.
But after I clicked the "spawn" button, it not redirected to the Jupyter Notebook interface. Just stay at the dialog screen...
I try to find some information in kubernetes and find it be killed as this issue...
So how to solve it? Anyone have some ideas? THANKS.

**Comment by pdmack** [comment]
Please report `kubectl describe pod jupyter-<your hub login name>`.

**Comment by myqq0000** [comment]
@pdmack Thank you for your timely reply. Sorry I don't know k8s well... I use this command to get report but it returns
`Error from server (NotFound): pods "jupyter-test" not found`
I think it because the pod be killed after it started few seconds.
I get some information from the namespace events like these:

> Killing container with id docker://notebook:Need to kill Pod
> kubelet ultron-pcn
> spec.containers{notebook}
> 1
> 2018-04-09T17:07 UTC
> 2018-04-09T17:07 UTC
> -
> Started container
> kubelet ultron-pcn
> spec.containers{notebook}
> 1
> 2018-04-09T17:06 UTC
> 2018-04-09T17:06 UTC
> -
> Created container
> kubelet ultron-pcn
> spec.containers{notebook}
> 1
> 2018-04-09T17:06 UTC
> 2018-04-09T17:06 UTC
> Container image "gcr.io/kubeflow-images-staging/tensorflow-1.7.0-notebook-cpu:v20180403-1f854c44" already present on machine
> kubelet ultron-pcn
> spec.containers{notebook}
> 1
> 2018-04-09T17:06 UTC
> 2018-04-09T17:06 UTC
> -
> MountVolume.SetUp succeeded for volume "no-api-access-please"
> kubelet ultron-pcn
> 1
> 2018-04-09T17:06 UTC
> 2018-04-09T17:06 UTC
> -
> Successfully assigned jupyter-test to ultron-pcn
> default-scheduler
> 1
> 2018-04-09T17:06 UTC
> 2018-04-09T17:06 UTC

I just know that the pod is killed.
But I can't find any useful information... 
I have tried the cpu and gpu versions (I have make the GPUs available for k8s.), but all meet the same issue.
Thank you for your reply~

**Comment by pdmack** [comment]
Can you read back to https://github.com/kubeflow/kubeflow/issues/554#issuecomment-378230674 above and confirm that your CPU actually supports AVX?

**Comment by myqq0000** [comment]
@pdmack Emmm... My CPU is i7-5930K. I think it supports AVX and I have checked it. 
And I don't use VM (e.g.minikube) to set up kubernetes so maybe don't have this problem.
Emmm... I'm new to k8s so don't know how to get more useful logs or other information.
If you have any more advice please tell me. I will try it as soon as possible.
Thanks a lot!

**Comment by pdmack** [comment]
Can you try https://github.com/kubeflow/kubeflow/issues/554#issuecomment-378062446 with whatever image is failing for you?

**Comment by pdmack** [comment]
Weird. Seems happy enough. I'll try here and get back to you.

**Comment by myqq0000** [comment]
@pdmack Sorry I am new to technology of container... 😅
I can visit the jupyter notebook with your advice.
But the kubeflow can't set up jupyter don't solve.
Thank you all the way~

**Comment by magreiner** [comment]
My jupyter pod also get terminated shortly after they are created (before the jupyter notebook is ready, maybe a timeout?). 

The container can be deployed manually in docker or kubernetes, without getting terminated (single-node kubernetes deployment).

The full description of the pod shortly before its killed:
Note the last line: "Killing container with id docker://notebook:Need to kill Pod"
[need-to-kill.log](https://github.com/kubeflow/kubeflow/files/1974959/need-to-kill.log)

However, modifying the restrictions (CPUs, Memory or GPUs) did not help.

Using the following image did solve this issue for me:
`gcr.io/kubeflow/tensorflow-notebook-gpu:8fbc341245695e482848ac3c2034a99f7c1e5763
`
The full comment from @yongwen can be found here:
https://github.com/kubeflow/kubeflow/issues/414#issuecomment-373192988

(The latest GPU-image did work but had issues with CUDA (libcublas.so.9.0: cannot open shared object file).

I'm happy to provide additional details if needed.

**Comment by ankushagarwal** [comment]
@magreiner 

Can you try `gcr.io/kubeflow-images-public/tensorflow-1.7.0-notebook-gpu:latest`

**Comment by magreiner** [comment]
@ankushagarwal The image `gcr.io/kubeflow-images-public/tensorflow-1.7.0-notebook-gpu:latest` does not get terminated!
But it does have some (probably unrelated) stability issues. If I run an example (https://colab.research.google.com/notebooks/gpu.ipynb) the python-kernel dies. Running the same code with the image from TensorFlow 1.5.1, it does not produce any errors.

@pdmack I'm not able to reproduce your error. Did you check the available disk space / free inodes?
https://github.com/moby/moby/issues/29508#issuecomment-267817461

Here my log:
`
 ~ > docker pull gcr.io/kubeflow-images-staging/tensorflow-1.5.1-notebook-gpu:v20180403-1f854c44
v20180403-1f854c44: Pulling from kubeflow-images-staging/tensorflow-1.5.1-notebook-gpu
22dc81ace0ea: Pull complete 
...
b5cb71f670e7: Pull complete 
Digest: sha256:a1755ea88a7820fbf3e2c11a4593a3d566ba22cc42ee7384c3b92b110ccd4025
Status: Downloaded newer image for gcr.io/kubeflow-images-staging/tensorflow-1.5.1-notebook-gpu:v20180403-1f854c44
Time: 0h:01m:49s    
`

**Comment by pdmack** [comment]
@magreiner I was able to pull other TF images _besides_ that one.
